src/_P021_Level.ino

```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+      relay_output = false; // Relay output state
+      break;
+    case P021_STATE_HEATING:
+      relay_output = true;  // Relay output state
+      break;
+    case P021_STATE_EXTEND:
+      relay_output = true;  // Relay output state
+      break;
+    case P021_STATE_FORCE:
+      relay_output = true;  // Relay output state
+      break;
+    default:                // unexpected state, switch pump off
+      relay_output = false; // Relay output state
       break;
-    }
   }
 
-  return success;
+  // Actuate the output pin taking output invert flag into account
+  if (validGpio(P021_GPIO_RELAY))
+  {
+    relay_output ^= P021_getConfigBit(P021_FLAGS, P021_INV_OUTPUT); // Invert when selected
+    digitalWrite(P021_GPIO_RELAY, relay_output ? HIGH : LOW);
+  }
+
+  #ifdef PLUGIN_021_DEBUG
+
+  if (loglevelActiveFor(LOG_LEVEL_DEBUG)) {
+    String log = F("P021: Calculated State= ");
+    log += P021_printControlState(new_control_state);
+    log += F(" -> ");
+    log += F("; relay= ");
+    log += relay_output;
+    log += F("; timer= ");
+    log += timePassedSince(timestamp) / 1000;
+    log += F("; mode= ");
+    log += P021_printControlMode(P021_OPMODE);
+    log += F("; value= ");
+    log += value;
+    log += F("; remote= ");
+    log += remote_state;
+    addLogMove(LOG_LEVEL_DEBUG, log);
+  }
+  #endif // ifdef PLUGIN_021_DEBUG
+
+  // Write back updated persistant control data:
+  // - The logical state of the output signal [on,OFF]
+  // - The state of the internal state machine
+  // - Timestamp for last change of the output signal
+  // Note: the actual state of the output is not stored, it can be calculated from the control_state
+  UserVar.setFloat(event->TaskIndex, P021_VALUE_OUTPUT, (float)(((P021_control_state)new_control_state == P021_STATE_IDLE) ? 0 : 1));
+  UserVar.setFloat(event->TaskIndex, P021_VALUE_STATE,  (float)new_control_state);
+  P021_timestamp[event->TaskIndex] = timestamp;
+
+  if (new_control_state != old_control_state)
+  {
+    sendData(event);
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Input evaluation to determine switching the control signal on
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+bool P021_check_on(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force)
+{
+  if (force)
+  {
+    return true;
+  }
+
+  if (symetric)
+  {
+    bool  isZero             = essentiallyZero(hysteresis);
+    float valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
+    float valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
+    return invert ? !definitelyGreaterThan(value, valueLowThreshold) : !definitelyLessThan(value, valueHighThreshold);
+  }
+  else
+  {
+    return invert ? (value < setpoint) : (value > setpoint);
```

Comment:
Why not also use `definitelyGreaterThan()/definitelyLessThan()` in the second return?

Reply:
I did not want to touch the existing algorithm. I learned that one shall not test for equal/unequal with floats due to inaccuracy errors; but comparing them is ok. I assume the operator will define a good hysteresis for the controller anyway as signal noise is probably a bigger issue. But if I am wrong here I can add the function.

Reply:
Well, using 2 different types of comparisons for variables of the same magnitude in a single function seems a bit weird, IMHO. The original code uses the `definitelyGreaterThan()/definitelyLessThan()` functions, so why not use them also for the added comparisons?

Reply:
That function can probably be simplified to:
```c++
bool P021_check_on(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force)
{
  if (force)
  {
    return true;
  }

  float valueLowThreshold  = setpoint;
  float valueHighThreshold = setpoint;

  if (symetric)
  {
    bool isZero = essentiallyZero(hysteresis);
    valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
    valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
  }
  return invert ? !definitelyGreaterThan(value, valueLowThreshold) : !definitelyLessThan(value, valueHighThreshold);
}
```

Reply:
I was thinking about the same style of simplification :-)

Reply:
I refactored it using the overview in the documentation.